File: StarcUp.UI/electron/src/services/core/NamedPipeProtocol.ts
/**
 * Named Pipe 통신 프로토콜 정의
 */

export enum MessageType {
  Request = 0,
  Response = 1, 
  Event = 2
}

/**
 * 기본 메시지 구조
 */
export interface BaseMessage {
  id: string
  type: MessageType
  timestamp: number
}

/**
 * 요청 메시지
 */
export interface RequestMessage extends BaseMessage {
  type: MessageType.Request
  command: string
  payload?: any
}

/**
 * 응답 메시지
 */
export interface ResponseMessage extends BaseMessage {
  type: MessageType.Response
  requestId: string
  success: boolean
  data?: any
  error?: string
}

/**
 * 이벤트 메시지
 */
export interface EventMessage extends BaseMessage {
  type: MessageType.Event
  event: string
  data?: any
}

/**
 * 모든 메시지 타입의 유니온
 */
export type NamedPipeMessage = RequestMessage | ResponseMessage | EventMessage

/**
 * 메시지 생성 헬퍼 함수들
 */
export class NamedPipeProtocol {
  
  /**
   * 고유 ID 생성
   */
  static generateId(): string {
    const timestamp = Date.now()
    const random = Math.random().toString(36).substring(2, 10)
    return `msg_${timestamp}_${random}`
  }

  /**
   * 요청 메시지 생성
   */
  static createRequest(command: string, payload?: any): RequestMessage {
    return {
      id: this.generateId(),
      type: MessageType.Request,
      timestamp: Date.now(),
      command,
      payload
    }
  }

  /**
   * 응답 메시지 생성
   */
  static createResponse(requestId: string, success: boolean, data?: any, error?: string): ResponseMessage {
    return {
      id: this.generateId(),
      type: MessageType.Response,
      timestamp: Date.now(),
      requestId,
      success,
      data,
      error
    }
  }

  /**
   * 이벤트 메시지 생성
   */
  static createEvent(event: string, data?: any): EventMessage {
    return {
      id: this.generateId(),
      type: MessageType.Event,
      timestamp: Date.now(),
      event,
      data
    }
  }

  /**
   * 메시지 타입 확인 헬퍼
   */
  static isRequest(message: any): message is RequestMessage {
    return message.type === MessageType.Request && typeof message.command === 'string'
  }

  static isResponse(message: any): message is ResponseMessage {
    return message.type === MessageType.Response && typeof message.requestId === 'string'
  }

  static isEvent(message: any): message is EventMessage {
    return message.type === MessageType.Event && typeof message.event === 'string'
  }
}

/**
 * 지원되는 명령 목록
 */
export const Commands = {
  Ping: 'ping',
  StartGameDetect: 'start-game-detect',
  StopGameDetect: 'stop-game-detect',
  GetGameStatus: 'get-game-status',
  GetUnitCounts: 'get-unit-counts',
  GetPlayerInfo: 'get-player-info',
  
  // 프리셋 관련 명령
  PresetInit: 'preset-init',
  PresetUpdate: 'preset-update'
} as const

/**
 * 지원되는 이벤트 목록
 */
export const Events = {
  GameDetected: 'game-detected',
  GameEnded: 'game-ended',
  GameStatusChanged: 'game-status-changed',
} as const

File: StarcUp.Core/Src/Infrastructure/Communication/NamedPipeProtocol.cs
using System;
using System.Text.Json;

namespace StarcUp.Infrastructure.Communication
{
    /// <summary>
    /// Named Pipe 통신 프로토콜 정의
    /// </summary>
    public static class NamedPipeProtocol
    {
        /// <summary>
        /// 메시지 타입
        /// </summary>
        public enum MessageType
        {
            Request,    // 요청
            Response,   // 응답
            Event       // 이벤트 (단방향)
        }

        /// <summary>
        /// 기본 메시지 구조
        /// </summary>
        public class BaseMessage
        {
            public string Id { get; set; }
            public MessageType Type { get; set; }
            public long Timestamp { get; set; }

            public BaseMessage()
            {
                Id = $"msg_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{Guid.NewGuid().ToString("N")[..8]}";
                Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            }
        }

        /// <summary>
        /// 요청 메시지
        /// </summary>
        public class RequestMessage : BaseMessage
        {
            public string Command { get; set; }
            public JsonElement? Payload { get; set; }

            public RequestMessage() : base()
            {
                Type = MessageType.Request;
            }

            public RequestMessage(string command, object payload = null) : this()
            {
                Command = command;
                if (payload != null)
                {
                    var json = JsonSerializer.Serialize(payload);
                    Payload = JsonDocument.Parse(json).RootElement;
                }
            }
        }

        /// <summary>
        /// 응답 메시지
        /// </summary>
        public class ResponseMessage : BaseMessage
        {
            public string RequestId { get; set; }
            public bool Success { get; set; }
            public JsonElement? Data { get; set; }
            public string Error { get; set; }

            public ResponseMessage() : base()
            {
                Type = MessageType.Response;
            }

            public ResponseMessage(string requestId, bool success, object data = null, string error = null) : this()
            {
                RequestId = requestId;
                Success = success;
                Error = error;
                
                if (data != null)
                {
                    var json = JsonSerializer.Serialize(data);
                    Data = JsonDocument.Parse(json).RootElement;
                }
            }
        }

        /// <summary>
        /// 이벤트 메시지
        /// </summary>
        public class EventMessage : BaseMessage
        {
            public string Event { get; set; }
            public JsonElement? Data { get; set; }

            public EventMessage() : base()
            {
                Type = MessageType.Event;
            }

            public EventMessage(string eventName, object data = null) : this()
            {
                Event = eventName;
                if (data != null)
                {
                    var json = JsonSerializer.Serialize(data);
                    Data = JsonDocument.Parse(json).RootElement;
                }
            }
        }

        /// <summary>
        /// 지원되는 명령 목록
        /// </summary>
        public static class Commands
        {
            public const string Ping = "ping";
            public const string StartGameDetect = "start-game-detect";
            public const string StopGameDetect = "stop-game-detect";
            public const string GetGameStatus = "get-game-status";
            public const string GetUnitCounts = "get-unit-counts";
            public const string GetPlayerInfo = "get-player-info";
            
            // 프리셋 관련 명령
            public const string PresetInit = "preset-init";
            public const string PresetUpdate = "preset-update";
        }

        /// <summary>
        /// 지원되는 이벤트 목록
        /// </summary>
        public static class Events
        {
            // 게임 프로세스 관련 이벤트
            public const string GameDetected = "game-detected";
            public const string GameEnded = "game-ended";
            public const string GameStatusChanged = "game-status-changed";
            
            // WorkerManager 관련 이벤트
            public const string WorkerStatusChanged = "worker-status-changed";
            public const string GasBuildingAlert = "gas-building-alert";
            public const string WorkerPresetChanged = "worker-preset-changed";
            
            // PopulationManager 관련 이벤트
            public const string SupplyAlert = "supply-alert";
            
            // UpgradeManager 관련 이벤트
            public const string UpgradeDataUpdated = "upgrade-data-updated";
            public const string UpgradeStateChanged = "upgrade-state-changed";
            public const string UpgradeCompleted = "upgrade-completed";
            public const string UpgradeCancelled = "upgrade-cancelled";
            public const string UpgradeInit = "upgrade-init";
        }
    }
}